src/shogun/regression/KRRNystrom.cpp

```diff
@@ -0,0 +1,128 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#include <cfloat>
+#include <shogun/regression/KRRNystrom.h>
+#include <shogun/labels/RegressionLabels.h>
+#include <shogun/mathematics/eigen3.h>
+#include <shogun/mathematics/Math.h>
+
+using namespace shogun;
+using namespace Eigen;
+
+CKRRNystrom::CKRRNystrom() : CKernelRidgeRegression()
+{
+	init();
+}
+
+CKRRNystrom::CKRRNystrom(float64_t tau, int32_t m, CKernel* k, CLabels* lab)
+: CKernelRidgeRegression(tau, k, lab)
+{
+	init();
+
+	m_num_rkhs_basis=m;
+}
+
+void CKRRNystrom::init()
+{
+	m_num_rkhs_basis=1000;
+}
+
+SGVector<int32_t> CKRRNystrom::subsample_indices()
+{
+	int32_t n=kernel->get_num_vec_lhs();
+	SGVector<int32_t> temp(n);
+	temp.range_fill();
+	CMath::permute(temp);
+	SGVector<int32_t> col(m_num_rkhs_basis);
+	for (auto i=0; i<m_num_rkhs_basis; ++i)
+		col[i]=temp[i];
+	CMath::qsort(col.vector, m_num_rkhs_basis);
+
+	return col;
+}
+
+bool CKRRNystrom::solve_krr_system()
+{
+	int32_t n=kernel->get_num_vec_lhs();
+
+	REQUIRE(m_num_rkhs_basis <= n, "Number of sampled rows/columns must be less than total\n");
+
+	SGVector<float64_t> y=((CRegressionLabels*)m_labels)->get_labels();
+	SGVector<int32_t> col=subsample_indices();
+	SGMatrix<float64_t> K_mm(m_num_rkhs_basis, m_num_rkhs_basis);
+	SGMatrix<float64_t> K_nm(n, m_num_rkhs_basis);
+	for (index_t i=0; i<m_num_rkhs_basis; ++i)
+	{
+		for (index_t j=0; j<m_num_rkhs_basis; ++j)
+			K_mm(i,j)=kernel->kernel(col[i],col[j]);
+		for (index_t j=0; j<n; ++j)
+			K_nm(j,i)=kernel->kernel(j,col[i]);
+	}
+
+	Map<MatrixXd> K_mm_eig(K_mm.matrix, m_num_rkhs_basis, m_num_rkhs_basis);
+	Map<MatrixXd> K_nm_eig(K_nm.matrix, n, m_num_rkhs_basis);
+	MatrixXd K_mn_eig = K_nm_eig.transpose();
+	Map<VectorXd> y_eig(y.vector, n);
+	VectorXd alphas_eig(m_num_rkhs_basis);
+
+	/* Calculate the Moore-Penrose pseudoinverse */
+	MatrixXd Kplus=K_mn_eig*K_nm_eig+m_tau*K_mm_eig;
+	SelfAdjointEigenSolver<MatrixXd> solver(Kplus);
+	if (solver.info()!=Success)
+	{
+		SG_WARNING("Eigendecomposition failed.")
+		return false;
+	}
+
+	/* Solve the system for alphas */
+	MatrixXd D=solver.eigenvalues().asDiagonal();
+	MatrixXd eigvec=solver.eigenvectors();
+	const float64_t tolerance=m_num_rkhs_basis*DBL_EPSILON*D.maxCoeff();
```

Comment:
I think this is (or should be) defined in CMath ... it is not portable this way. Might be a reason for an error